StrataTest/Backends/CBMC/BoogieToCProverGOTO.lean

```diff
@@ -0,0 +1,235 @@
+    let new := name_alt.getD (Boogie.BoogieIdent.toPretty name)
+    .init new ty e' .empty
+  | .set name e _ =>
+    let e' := substVarNames Boogie.BoogieIdent.toPretty e frto
+    let name_alt := frto.find? (Boogie.BoogieIdent.toPretty name)
+    let new := name_alt.getD (Boogie.BoogieIdent.toPretty name)
+    .set new e' .empty
+  | .havoc name _ =>
+    let name_alt := frto.find? (Boogie.BoogieIdent.toPretty name)
+    let new := name_alt.getD (Boogie.BoogieIdent.toPretty name)
+    .havoc new .empty
+  | .assume label e _ =>
+    let e' := substVarNames Boogie.BoogieIdent.toPretty e frto
+    .assume label e' .empty
+  | .assert label e _ =>
+    let e' := substVarNames Boogie.BoogieIdent.toPretty e frto
+    .assert label e' .empty
+
+def Boogie.Cmds.renameVars (frto : Map String String)
+    (cs : Imperative.Cmds Boogie.Expression) : Imperative.Cmds Boogie.ExprStr :=
+  match cs with
+  | [] => []
+  | c :: crest => [(Boogie.Cmd.renameVars frto c)] ++ (renameVars frto crest)
+
+-------------------------------------------------------------------------------
+
+structure CProverGOTO.Context where
+  program : CProverGOTO.Program
+  locals : List String
+  formals : Map String CProverGOTO.Ty
+  ret : CProverGOTO.Ty
+
+structure CProverGOTO.Json where
+  symtab : Lean.Json := .null
+  goto   : Lean.Json := .null
+
+open Strata in
+def CProverGOTO.Context.toJson (programName : String) (ctx : CProverGOTO.Context) :
+  CProverGOTO.Json :=
+  let fn_symbol : Map String CProverGOTO.CBMCSymbol :=
+    [CProverGOTO.createFunctionSymbol programName ctx.formals ctx.ret]
+  let formals : Map String CProverGOTO.CBMCSymbol :=
+    ctx.formals.map (fun (name, ty) =>
+        CProverGOTO.createGOTOSymbol programName name (CProverGOTO.mkFormalSymbol programName name)
+          (isParameter := true) (isStateVar := true) (ty := some ty))
+  let locals : Map String CProverGOTO.CBMCSymbol :=
+    ctx.locals.map (fun name =>
+        CProverGOTO.createGOTOSymbol programName name (CProverGOTO.mkLocalSymbol programName name)
+          (isParameter := false) (isStateVar := false) (ty := none))
+  let symbols := Lean.toJson (fn_symbol ++ formals ++ locals)
+  let goto_functions := CProverGOTO.programsToJson [(programName, ctx.program)]
+  { symtab := symbols, goto := goto_functions }
+
+open Lambda.LTy.Syntax in
+def transformToGoto (boogie : Boogie.Program) : Except Format CProverGOTO.Context := do
+  let T := { Lambda.TEnv.default with functions := Boogie.Factory, knownTypes := Boogie.KnownTypes }
+  let (boogie, _T) ← Boogie.Program.typeCheck T boogie
+  dbg_trace f!"[Strata.Boogie] Type Checking Succeeded!"
+  if h : boogie.decls.length = 1 then
+    let decl := boogie.decls[0]'(by exact Nat.lt_of_sub_eq_succ h)
+    match decl.getProc? with
+    | none => .error f!"[transformToGoto] We can process only Boogie procedures at this time. \
+                        Declaration encountered: {decl}"
+    | some p =>
```

Comment:
It's not critical for this PR, as I know all this code is going to undergo a lot more refactoring soon, but let's break this branch of the match out into a `transformProcedureToGoto` function (and maybe rename this one to `transformProgramToGoto`.